### PR TITLE
[ENH] Add Kernel-based Conditional Independence (KCI) test (#3106)            

### PR DESCRIPTION
**Pull request by hanara2112**
# The following checklist is mandatory.

Your PR will be closed if you remove the checklist or do not answer the questions to a satisfactory level. Use of LLMs is **strictly forbidden** for any part of this checklist (including for improving language), and will result in a **ban** if we find any use of LLMs.

### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md)?
- [x] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft.
- [x] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

Please answer the following questions:

- Did you use an LLM for any assistance with this PR? Please describe in detail how and what you used it for?                                        
  - Yes, mainly for scaffolding code and tests. I verified the logic myself. The bandwidth part I figured out by reading causal-learn since the LLM got it wrong.
  
- Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.   
  - Yes. It uses RBF kernels + HSIC for independence. For conditional, Z is regressed out before computing the statistic. Main fix was correcting bandwidth logic for HSIC vs KCI.                                                              

- What steps have you taken to verify that the changes correctly address the 
  issue? And what edge cases have you considered? Other than running tests,   
  what else have you verified?        
  - Matched stats and p-values with causal-learn across a few scenarios. Also checked small samples and edge cases.                                      

- Has the LLM added try-except blocks? They will need to be removed; any error handling must be explicit.                                                 
  - No
- Have you used LLM for generating tests? They need to be compressed into a 
  smaller number of tests without reducing coverage.         
  - Initially yes, but I simplified and reduced them.               

                                                                              
Issue number(s) that this pull request fixes
  - Solves #3106 issue part of the umbrella issue #3099 

- List of changes to the codebase in this pull request                        
   - pgmpy/ci_tests/kci.py : new addition
   - pgmpy/tests/test_ci_tests/test_kci.py: new addition

### CODE REVIEW

File: pgmpy/tests/test_ci_tests/test_kci.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import unittest
```
Comment (Jatinbhardwaj-093) [inline pgmpy/tests/test_ci_tests/test_kci.py]: Use pytest

File: pgmpy/ci_tests/kci.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+        The dataset in which to test the independence condition.
+
+    kernel_X : str, default="rbf"
+        Kernel for variable X. Currently only ``"rbf"`` is supported.
+
+    kernel_Y : str, default="rbf"
+        Kernel for variable Y. Currently only ``"rbf"`` is supported.
+
+    kernel_Z : str, default="rbf"
+        Kernel for conditioning variables Z. Currently only ``"rbf"`` is supported.
```
Comment (ankurankan) [inline pgmpy/ci_tests/kci.py]: These should be sklearn Kernel instances. This will allow the users to have finer control over the kernels and also be able to choose any kernel they want. We can keep the default as sklearn's RBF kernel.

File: pgmpy/ci_tests/kci.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+
+    width_X : float or None, default=None
+        Bandwidth (sigma) for the X kernel. If ``None``, estimated from the data
+        using a sample-size-based heuristic.
+
+    width_Y : float or None, default=None
+        Bandwidth (sigma) for the Y kernel. If ``None``, estimated from the data.
+
+    width_Z : float or None, default=None
+        Bandwidth (sigma) for the Z kernel. If ``None``, estimated from the data.
```
Comment (ankurankan) [inline pgmpy/ci_tests/kci.py]: These would get removed after the above change.

File: pgmpy/ci_tests/kci.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+    regularization : float, default=1e-3
+        Regularization parameter :math:`\epsilon` used in the conditional test to
+        invert the kernel matrix on :math:`Z`.
+
```
Comment (ankurankan) [inline pgmpy/ci_tests/kci.py]: Is this a standard parameter of the test? This feels something just for numerical stability?

File: pgmpy/ci_tests/kci.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    Examples
+    --------
+    >>> import numpy as np
+    >>> import pandas as pd
+    >>> from pgmpy.ci_tests import KCI
+    >>> rng = np.random.default_rng(seed=42)
+    >>> data = pd.DataFrame(rng.standard_normal((300, 3)), columns=["X", "Y", "Z"])
+    >>> test = KCI(data=data)
+    >>> test("X", "Y", [], significance_level=0.05)
+    True
```
Comment (ankurankan) [inline pgmpy/ci_tests/kci.py:111]: Show the attributes as well.

File: pgmpy/ci_tests/kci.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        self.kernel_Y = kernel_Y
+        self.kernel_Z = kernel_Z
+        self.width_X = width_X
+        self.width_Y = width_Y
+        self.width_Z = width_Z
+        self.regularization = regularization
+        super().__init__()
+
+    @staticmethod
+    def _rbf_kernel(X, width):
```
Comment (ankurankan) [inline pgmpy/ci_tests/kci.py]: Can be removed after accepting sklearn kernels

File: pgmpy/estimators/CITests.py [hunk 7]
```diff
@@ -939,3 +939,65 @@ def pearsonr_equivalence(X, Y, Z, data, boolean=True, delta_threshold=0.1, **kwa
         return p_value < kwargs.get("significance_level", 0.05)
     else:
         return coeff, p_value
+
+
+@ci_registry.register(name="kci", data_types=["continuous"])
```
Comment (ankurankan) [inline pgmpy/estimators/CITests.py]: No need to add here. This will be deprecated in the next couple of releases.

File: pgmpy/tests/test_ci_tests/test_kci.py (hunk 1 above)
Comment (ankurankan) [inline pgmpy/tests/test_ci_tests/test_kci.py]: And add tests for comparing values with other packages.

**Review by ankurankan — CHANGES_REQUESTED** [review]
Had a design level review and have added some comments.

**Comment by hanara2112** [comment]
Hi @ankurankan, I went through the KCI implementation again to make sure I properly understand it before pushing more changes. Writing down my current understanding + a few questions, please correct me if anything is off.

Currently the implementation splits into two cases depending on whether Z is empty or not.

When Z is empty (HSIC case):

1. Normalize X and Y
2. Compute RBF kernels
3. Center the kernels
4. Compute the statistic using element-wise product of centered kernels
5. Approximate p-value using gamma fit (based on traces / Frobenius norms)

When Z is present (conditional case):

1. Compute kernels for X, Y, Z
2. For X, use augmented input [X, 0.5 * Z]
3. Center all kernels
4. Residualize using something like ε * pinv(Kz + εI)
5. Compute statistic similarly
6. Estimate moments via eigendecomposition

This might be a bit rough for the conditional case, but this is how I’m currently reading the flow.

**Comment by hanara2112** [comment]
Things i am unsure about and would need discussion:

1. Bandwidth handling: when user passes a kernel _estimate_bandwidth only runs when kernel=None.
If a user passes RBF() directly, it uses default length_scale=1.0, which might not match the median heuristic on normalized data.
Would it be better to: just document this clearly, or expose _estimate_bandwidth so users can compute it before passing a kernel?

2. Augmenting X with 0.5 * Z
This follows the paper / causal-learn, but when a custom kernel_X is passed, the input changes from X → [X, Z].
Should this just be documented, or is there a cleaner way to handle this?

3. Complexity concerns
Since this uses pseudoinverse / eigendecomposition, it looks like ~O(n³) time and O(n²) memory.
Do you think it’s worth adding a warning/check for large n, or is that out of scope for now?

File: pgmpy/tests/test_ci_tests/test_kci.py (hunk 1 above)
Comment (hanara2112) [reply pgmpy/tests/test_ci_tests/test_kci.py]: other CI tests use unit test. can you please clarify?

File: pgmpy/ci_tests/kci.py (hunk 4 above)
Comment (hanara2112) [reply pgmpy/ci_tests/kci.py]: yes, it's for numerical stability. it's the ε in ε(Kz + εI)⁻¹. 
The causal-learn implementation hardcodes it to 1e-3 as well. I've removed it as a user-facing parameter in the latest refactor. It's now just an internal constant in _conditional_test.

File: pgmpy/tests/test_ci_tests/test_kci.py (hunk 1 above)
Comment (hanara2112) [reply pgmpy/tests/test_ci_tests/test_kci.py]: Added TestKCICompareCausalLearn that compares pgmpy's results against causal-learn's KCI on 4 scenarios       
(unconditional independent/dependent, conditional independent/dependent). Both implementations agree on all cases.

**Comment by hanara2112** [comment]
I revisited the implementation and realized my conditional projection step was incorrect. I wasn’t using the proper residual-forming operator R_z = I - K_z (K_z + \epsilon I)^{-1}.
I’m updating this along with correcting HSIC scaling and reviewing the gamma approximation to better match the original paper.

File: pgmpy/tests/test_ci_tests/test_kci.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    def test_get_ci_test_factory(self):
+        test = get_ci_test(test="kci", data=self.df_ind)
+        self.assertIsInstance(test, KCI)
+
+
+@unittest.skipUnless(
+    __import__("importlib").util.find_spec("causallearn"),
+    "causallearn not installed",
+)
+class TestKCICompareCausalLearn(unittest.TestCase):
```
Comment (ankurankan) [inline pgmpy/tests/test_ci_tests/test_kci.py]: We should not just look at whether values are greater than or less than 0.05. We need to compare exactly the statistic and p-values computed by us vs what causallearn's implementation returns. That will help identify any subtle differences in the implementation. Also, please don't compute the values from causallearn in the tests, as this will add a dependency for our package:

- You should run the causallearn tests on your local machine and copy the outputted values in the tests here.
- Add the causal learn code used to generate the values as comment in the code, so that we can reproduce it later if required.

File: pgmpy/tests/test_ci_tests/test_kci.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
+import unittest
+
+import numpy as np
+import pandas as pd
+from sklearn.gaussian_process.kernels import RBF
+
+from pgmpy.ci_tests import KCI, get_ci_test
+
+
+class TestKCI(unittest.TestCase):
```
Comment (ankurankan) [inline pgmpy/tests/test_ci_tests/test_kci.py]: Tests need to be compressed. These look LLM generated.

**Comment by ankurankan** [comment]
> Would it be better to: just document this clearly, or expose _estimate_bandwidth so users can compute it before passing a kernel?

This can be a boolean argument in the constructor. If yes, we can use sklearn's `set_params` method to update the Kernal constructors. This behavior should be documented clearly.

 > Augmenting X with 0.5 * Z
This follows the paper / causal-learn, but when a custom kernel_X is passed, the input changes from X → [X, Z].
Should this just be documented, or is there a cleaner way to handle this?

Can you point me to where exactly is this in the paper? 

> Since this uses pseudoinverse / eigendecomposition, it looks like ~O(n³) time and O(n²) memory.
Do you think it’s worth adding a warning/check for large n, or is that out of scope for now?

This is fine. It's a very well known issue of KCI; RCIT improves on this.

File: pgmpy/ci_tests/kci.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+
+
+class KCI(_BaseCITest):
+    r"""
+    Kernel-based Conditional Independence (KCI) test [1].
+
+    When :math:`Z = \emptyset`, this reduces to the HSIC (Hilbert-Schmidt Independence
+    Criterion) test for unconditional independence. When :math:`Z \neq \emptyset`, it
+    performs a kernel-based conditional independence test (KCIT) by regressing out the
+    effect of :math:`Z` from the kernel matrices of :math:`X` and :math:`Y`.
```
Comment (ankurankan) [inline pgmpy/ci_tests/kci.py]: Try to add the exact equations of what we are computing in the test. Refer to other tests for example.

**Comment by ankurankan** [comment]
@hanara2112 Did you use an LLM to write the PR checklist?

File: pgmpy/ci_tests/kci.py (hunk 4 above)
Comment (hanara2112) [reply pgmpy/ci_tests/kci.py]: should this be included in the parameter docstring?

Comment (ankurankan) [reply pgmpy/ci_tests/kci.py]: It can stay as an argument but its purpose muct be documented properly.

**Comment by hanara2112** [comment]
Hi @ankurankan, while reading the paper and doing some finals check for the implementation i noticed that in Section 3.5 (page 6 from the KCI paper) , the authors suggest to use half the kernel width for K_z compared to what's used for Kx and Ky. 

In our current implementation and in causal-learn, all 3 kernels share the same bandwidth from empirical_width_kci. 
Should we keep it this way to stay consistent with causal-learn, or should we follow the paper and use ls / 2 for the Z kernel?

**Comment by hanara2112** [comment]
> > Would it be better to: just document this clearly, or expose _estimate_bandwidth so users can compute it before passing a kernel?

> This can be a boolean argument in the constructor. If yes, we can use sklearn's `set_params` method to update the Kernal constructors. This behavior should be documented clearly.

Right now, if a custom kernel_X is passed, it’s used as is, ie. no bandwidth estimation is applied. Empirical bandwidth is only used when the kernel is None.

We could introduce an auto_bandwidth flag (default True) with this behavior:

  1. auto_bandwidth=True: estimate bandwidth via _empirical_width_kci(Z) and override length_scale using set_params (for compatible kernels like RBF).
  2. auto_bandwidth=False (or no Z): use the kernel exactly as provided.

Although one important thing to notice here is that this only works for kernels with a length_scale parameter. For custom kernels, we can either:

  - restrict support to such kernels, or
  - use a try/except and document the requirement.


> > Augmenting X with 0.5 * Z
> > This follows the paper / causal-learn, but when a custom kernel_X is passed, the input changes from X → [X, Z].
> > Should this just be documented, or is there a cleaner way to handle this?
> 
> Can you point me to where exactly is this in the paper?

The 0.5Z augmentation isn't actually from the paper but from the casual-learn codebase. The paper defines X" = (X,Z) with a product kernel i.e. two seperate kernels multiplied element-wise. The concatenation of X with 0.5Z usinga single RBF is a casual-learn design choice for easier reference value comparison.

Since the paper uses separate kernels for X and Z, a cleaner approach matching the paper would be to compute kernel_X(X) * kernel_Z(Z) element-wise for Kx" rather than feeding [X, 0.5*Z] into kernel_X. That way a custom kernel_X would operate on X alone, as the user would expect. Should I refactor it this way, or keep it consistent with causal-learn's concatenation approach ? but that would probably change the output values.

**Comment by hanara2112** [comment]
> @hanara2112 Did you use an LLM to write the PR checklist?

that was a intial checklist while starting to code. i will change it and write it by myself now since the kci implementation seems close to completion after a review. 
Apologies for any inconvienience.

File: pgmpy/tests/test_ci_tests/test_kci.py (hunk 8 above)
Comment (hanara2112) [reply pgmpy/tests/test_ci_tests/test_kci.py]: Done. TestKCICompareCausalLearn now directly checks stats and p-values against hardcoded causal-learn reference values (generated locally from v0.1.4.5), so there’s no external dependency. The docstring aslo includes the full reproduction code. After fixing the bandwidth issue, pgmpy matches causal-learn exactly across the test cases.

**Comment by hanara2112** [comment]
Hi @ankurankan , this PR is ready for review. Can you please let me know if any changes required ?

**Comment by hanara2112** [comment]
Working on this!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by hanara2112** [comment]
superseded by https://github.com/pgmpy/pgmpy/pull/3297.
